### PR TITLE
support streaming in language feature providers (Observable returns)

### PR DESCRIPTION
**Pull request by sqs**
Sourcegraph extensions can now return Observables from `provideHover`, `provideReferences`, and other language feature provider methods.

This is useful for:

- Streaming references, especially for external references, which take a lot longer than local references.
- Streaming hovers while the workspace is initializing, so the user doesn't need to keep moving their mouse around on tokens to see if it's ready yet.

Demo videos:

- https://slack-files.com/T02FSM7DL-FE6RZTVJ6-24bbd82b96
- https://slack-files.com/T02FSM7DL-FE6MR49JQ-6169e33377

Example code:

```typescript
import * as sourcegraph from 'sourcegraph'
import { interval } from 'rxjs'
import { map, bufferCount } from 'rxjs/operators'

export function activate(): void {
    // Normal (non-streaming) provider
    sourcegraph.languages.registerHoverProvider(['*'], {
        provideHover: () => ({
            contents: { value: ' Hello from an extension!' },
        }),
    })

    // "Streaming" hover provider
    sourcegraph.languages.registerHoverProvider(['*'], {
        provideHover: () =>
            interval(100).pipe(
                map(n => ({
                    contents: { kind: sourcegraph.MarkupKind.Markdown, value: `Counter: ${n}` },
                }))
            ),
    })

    // "Streaming" references provider
    sourcegraph.languages.registerReferenceProvider(['*'], {
        provideReferences: doc =>
            interval(500).pipe(
                bufferCount(7),
                map(values => {
                    const lines = doc.text.split('\n')
                    const lineNumbers = values.map(value => (value * 7) % lines.length)
                    return lineNumbers.filter(line => lines[line].length > 3).map(
                        line =>
                            ({
                                uri: new sourcegraph.URI(doc.uri),
                                range: new sourcegraph.Range(
                                    line,
                                    lines[line].length / 3,
                                    line,
                                    (lines[line].length * 2) / 3
                                ),
                            } as sourcegraph.Location)
                    )
                })
            ),
    })
}
```

TODOs (post-merge):

- Further cleanup in the jsonrpc2 TS directory. We are no longer using standard JSON-RPC 2.0, so we should remove other JSON-RPC 2.0 features we don't use and update docs.

### CODE REVIEW

**Comment by sqs** [comment]
@chrismwendt 	This does not directly address external references yet. I need to think more about how the API for those should look. I am leaning toward not differentiating between the two, on the principle that the distinction between local and external references is arbitrary (local references inside of a monorepo are effectively external references, at least as it concerns many use cases for needing references).

cc @francisschmaltz will chat with you for help here (and to revisit the reference panel designs you made, which are the right direction)

File: shared/src/api/protocol/jsonrpc2/connection.ts [hunk 1]
```diff
@@ -433,22 +454,22 @@ function _createConnection(transports: MessageTransports, logger: Logger, strate
             }
         } else {
             const key = String(responseMessage.id)
-            const responsePromise = responsePromises[key]
+            const responsePromise = responseObservables[key]
```
Comment (chrismwendt) [inline shared/src/api/protocol/jsonrpc2/connection.ts]: `s/responsePromise/responseObservable/`?

File: shared/src/api/protocol/jsonrpc2/connection.ts [hunk 2]
```diff
@@ -189,16 +200,13 @@ function _createConnection(transports: MessageTransports, logger: Logger, strate
         if (isRequestMessage(message)) {
             queue.set(createRequestQueueKey(message.id), message)
         } else if (isResponseMessage(message)) {
-            queue.set(createResponseQueueKey(message.id), message)
+            const key = createResponseQueueKey(message.id) + Math.random() // TODO(sqs)
```
[the remark this answers is not in the document]
Comment (sqs) [reply shared/src/api/protocol/jsonrpc2/connection.ts]: It's to make responses with complete:false not all use the same key space. It's a hack that I will clean up.

**Comment by sqs** [comment]
@chrismwendt:

> Does that accurately reflect what's happening?



Yep

**Pull request merged**